[Alerting]: Extend quota service to optionally set limits on alerts

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
- Extends quota service to optionally set limits on alerts
- Checks alert rule quota limit when POSTing new alert rules (via the API)

~The `alert_rule` table is created behind a feature flag so I have to fix and return the alert quota only if this is enabled.~ fixed

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:
start grafana with quota enabled and zero alert rule limit:
`GF_QUOTA_ORG_ALERT_RULE=0 GF_QUOTA_ENABLED=true GF_FEATURE_TOGGLES_ENABLE=ngalert GF_LOG_LEVEL=debug  make run`

get quota:
```
curl http://admin:admin@localhost:3000/api/orgs/1/quotas |jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   275  100   275    0     0  10576      0 --:--:-- --:--:-- --:--:-- 10576
[
  {
    "org_id": 1,
    "target": "org_user",
    "limit": 10,
    "used": 2
  },
  {
    "org_id": 1,
    "target": "data_source",
    "limit": 10,
    "used": 35
  },
  {
    "org_id": 1,
    "target": "dashboard",
    "limit": 100,
    "used": 75
  },
  {
    "org_id": 1,
    "target": "api_key",
    "limit": 10,
    "used": 0
  },
  {
    "org_id": 1,
    "target": "alert_rule",
    "limit": 100,
    "used": 4
  }
]
```

```
curl -XPOST http://admin:admin@localhost:3000/api/ruler/$grafanaRecipient/api/v1/rules/$namespace -d '
> {
>   "name": "group42",
>   "interval": "10s",
>   "rules": [
>     {
>       "grafana_alert": {
>         "title": "reduced testdata query",
>         "condition": "B",
>         "data": [
>           {
>             "refId": "A",
>             "queryType": "",
>             "relativeTimeRange": {
>               "from": 18000,
>               "to": 10800
>             },
>             "model": {
>               "alias": "just-testing",
>               "datasource": "gdev-testdata",
>               "datasourceUid": "000000004",
>               "intervalMs": 1000,
>               "maxDataPoints": 100,
>               "orgId": 0,
>               "refId": "A",
>               "scenarioId": "csv_metric_values",
>               "stringInput": "1,20,90,30,5,0"
>             }
>           },
>           {
>             "refId": "B",
>             "queryType": "",
>             "relativeTimeRange": {
>               "from": 18000,
>               "to": 10800
>             },
>             "model": {
>               "datasource": "__expr__",
>               "datasourceUid": "__expr__",
>               "expression": "$A",
>               "intervalMs": 2000,
>               "maxDataPoints": 200,
>               "orgId": 0,
>               "reducer": "mean",
>               "refId": "B",
>               "type": "reduce"
>             }
>           }
>         ],
>         "no_data_state": "NoData",
>         "exec_err_state": "Alerting"
>       }
>     }
>   ]
> ' |jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1399  100    27  100  1372   1350  68600 --:--:-- --:--:-- --:--:-- 69950
{
  "message": "quota reached"
}
```

I have modified the testinfra for making authenticated requests.

papagian committed:
Quota: Extend service to set limiit on alerts

papagian committed:
Add test for applying quota to alert rules

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Fix tests

oddlittlebird commented:
This needs a milestone.

oddlittlebird commented on docs/sources/administration/configuration.md:
```diff
@@ -1020,6 +1020,9 @@ Limit the number of data sources allowed per organization. Default is 10.
 Limit the number of API keys that can be entered per organization. Default is 10.
+### org_alert_rulle
+
+Limit the number of alert rulles that can be entered per organization. Default is 10.
```
```suggestion
Limit the number of alert rules that can be entered per organization. Default is 10.
```

oddlittlebird reviewed (commented):
A few minor doc suggestions

achatterjee-grafana commented on docs/sources/administration/configuration.md:
```diff
@@ -1020,6 +1020,9 @@ Limit the number of data sources allowed per organization. Default is 10.
 ### org_api_key
 
 Limit the number of API keys that can be entered per organization. Default is 10.
+### org_alert_rulle
```
org_alert_rule?

achatterjee-grafana reviewed (changes requested):
Commented on a possible typo.

achatterjee-grafana commented:
Is this change for V8.0 and part of the new alerting feature?

papagian commented:
> Is this change for V8.0 and part of the new alerting feature?

yes, I've added the milestone

papagian committed:
Apply suggestions from code review

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Get used alert quota only if naglert is enabled

papagian committed:
Fix docs

papagian committed:
Use contant

papagian committed:
Reorganise tests

papagian committed:
Set alert limit to zero if nglalert is not enabled

papagian committed:
lint

papagian committed:
Fix test

papagian commented:
I have fixed the issues; it's now ready for review

kylebrandt reviewed (approved):
LGTM

wbrowne commented on pkg/services/sqlstore/quota.go:
```diff
@@ -78,17 +84,21 @@ func GetOrgQuotas(query *models.GetOrgQuotasQuery) error {
-		if err := x.SQL(rawSQL, q.OrgId).Find(&resp); err != nil {
-			return err
+		var used int64
+		if q.Target != "alert_rule" || query.IsNgAlertEnabled {
```
Looks like you're missing the use of the `ALERT_RULE_TARGET` const here and also line 199 😄

wbrowne commented on pkg/middleware/quota_test.go:
```diff
@@ -208,6 +208,61 @@ func TestMiddlewareQuota(t *testing.T) {
+		})
+
+		middlewareScenario(t, "org alert quota reached but ngalert disabled", func(t *testing.T, sc *scenarioContext) {
+			// this schenario can only happen if the feature was enabled and later disabled
```
```suggestion
			// this scenario can only happen if the feature was enabled and later disabled
```

wbrowne commented on pkg/services/quota/quota.go:
```diff
@@ -151,6 +151,12 @@ func (qs *QuotaService) getQuotaScopes(target string) ([]models.QuotaScope, erro
 			models.QuotaScope{Name: "global", Target: target, DefaultLimit: qs.Cfg.Quota.Global.Session},
 		)
 		return scopes, nil
+	case "alert_rule": // target need to much the respective database name
```
Is this comment necessary to keep? I guess it should say "match" in case it should stay

wbrowne commented on pkg/services/sqlstore/quota_test.go:
```diff
@@ -87,12 +89,19 @@ func TestQuotaCommandsAndQueries(t *testing.T) {
 				So(err, ShouldBeNil)
 				So(query.Result.Used, ShouldEqual, 0)
 			})
+			Convey("Should be able to get used zero used org alert quota when table does not exist (ngalert is not enabled - default case)", func() {
```
```suggestion
			Convey("Should be able to get zero used org alert quota when table does not exist (ngalert is not enabled - default case)", func() {
```

papagian committed:
Apply suggestions from code review

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Fix prometheus test

It used to fail because the folder was created under organisation
with ID 1 (main org.) instead of 2 (the anonymous organisation)

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

wbrowne reviewed (approved):
Looks good AFAICT!

The only thing I wonder is what the impact is on build time due to more individual integration tests (a general problem) - I haven't done any profiling really, but am curious if in some cases we'd be better to reduce re-creating the database and server. Just some food for thought for now!

papagian commented:
> Looks good AFAICT!
> 
> The only thing I wonder is what the impact is on build time due to more individual integration tests (a general problem) - I haven't done any profiling really, but am curious if in some cases we'd be better to reduce re-creating the database and server. Just some food for thought for now!

good point! The integration tests were indeed slow because of re-creation of the database but [we have changed it](https://github.com/grafana/grafana/pull/28265) and we just truncate the tables instead following [this](https://github.com/grafana/grafana/pull/26858#issuecomment-692661118) suggestion.

papagian committed:
Merge remote-tracking branch 'origin/master' into papagian/alerting-apply-quota

papagian committed:
Fix failing test

createFolder() used to create the folder under the anonymous
organisation. This has changed so the requests has to be authenticated.
